### PR TITLE
Implement CartCommands and increase test coverage

### PR DESCRIPTION
**Pull request by teo-jun-xiong**
Features
- Abstract out `CartAddCommand`'s code to `CartAddRecipeIngredientCommand` and `CartAddIngredientCommand` respectively 
- Implement `CartAddRecipeIngredientCommand#execute()`
- Override `UniqueIngredientList#toString()`

Testing
- Add `CartAddIngredientCommandTest`
- Add `CartClearCommandTest`
- Add `CartRemoveIngredientCommandTest`
- Add `CookbookAddCommandTest`
- Add `CookbookRemoveCommandTest`
- Add `CookbookViewCommandTest`
- Add `InventoryAddIngredientCommandTest`
- Add `InventoryRemoveIngredientCommandTest`
- Add `RecipeAddIngredientCommandTest`
- Add `RecipeAddStepCommandTest`
- Add `RecipeAddTagCommandTest`
- Add `RecipeRemoveIngredientCommandTest`
- Add `RecipeRemoveStepCommandTest`
- Add `RecipeRemoveTagCommandTest`
- Add `CartAddCommandParserTest`
- Add `CartAddIngredientCommandParserTest`
- Add `CartClearCommandParserTest`
- Add `CartCommandParserTest`
- Add `CartRemoveIngredientCommandParserTest`
- Add `CookbookAddCommandParserTest`
- Add `CookbookRemoveCommandParserTest`
- Add `CookbookViewCommandParserTest`
- Add `CookbookCommandParserTest`
- Add `InventoryAddIngredientCommandParserTest`
- Add `InventoryRemoveIngredientCommandParserTest`
- Add `InventoryCommandParserTest`
- Add `RecipeAddCommandParserTest`
- Add `RecipeRemoveCommandParserTest`
- Add `RecipeCommandParserTest`
- Implement `RecipeNameContainsKeywordsPredicateTest`
- Implement `RecipeTest`

- **BUG**: added ingredients/tags/steps somehow persists in other tests, causing them to fail if all 476 tests are executed. Temporarily fixed by using a different recipe from `TypicalRecipes`.

### CODE REVIEW

File: src/main/java/seedu/address/model/IngredientList.java [hunk 1]
```diff
@@ -94,4 +94,8 @@ public CompatibleIngredientList getCompatibleIngredientList() {
     public ObservableList<Ingredient> getIngredientList() {
         return ingredients.asUnmodifiableObservableList();
     }
+
+    public void removeAllCartIngredients() {
+        ingredients.removeAllIngredients();
+    }
```
Comment (gohkh) [inline src/main/java/seedu/address/model/IngredientList.java]: It seems like this method should probably be in the `Cart` class rather that the `IngredientList` class which `Inventory` also extends from?

File: src/main/java/seedu/address/model/Model.java [hunk 2]
```diff
@@ -248,5 +248,5 @@
      */
     void updateFilteredCartIngredientList(Predicate<Ingredient> predicate);
 
-    void removeAllCartIngredient();
+    void removeAllCartIngredients();
```
Comment (gohkh) [inline src/main/java/seedu/address/model/Model.java]: I just went to take a look at AB3, the clear command is executed by using `model.setAddressBook(new AddressBook())`. I think we could do the same for clearing the cart rather than adding this new method, or otherwise we should add this method for the cookbook and inventory too. I didn't notice it in the original PR implementing the cart clear command.

File: src/main/java/seedu/address/model/IngredientList.java (hunk 1 above)
Comment (teo-jun-xiong) [reply src/main/java/seedu/address/model/IngredientList.java]: Hmm, i think we could rename it and keep it in IngredientList, and maybe it could be repurposed?

File: src/main/java/seedu/address/model/Model.java (hunk 2 above)
Comment (teo-jun-xiong) [reply src/main/java/seedu/address/model/Model.java]: alright, i'll revert to using `set`

**Review by hans555 — APPROVED** [review]

File: src/main/java/seedu/address/logic/parser/cart/CartAddRecipeIngredientCommandParser.java [hunk 3]
```diff
@@ -21,19 +21,31 @@
...
+                    CartAddCommand.MESSAGE_USAGE));
+        }
+
         try {
+            args = removeRecipePrefix(args);
             recipeNumber = Integer.parseInt(args);
-        } catch (NumberFormatException ne) {
-            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT,
+        } catch (NumberFormatException e) {
+            throw new NumberFormatException(String.format(MESSAGE_INVALID_COMMAND_FORMAT,
```
Comment (gohkh) [inline src/main/java/seedu/address/logic/parser/cart/CartAddRecipeIngredientCommandParser.java]: Should this be throwing `ParseException` instead? I don't think the caller of the method should know what the original exception was, other than through the error message.

File: src/main/java/seedu/address/logic/parser/cart/CartRemoveIngredientCommandParser.java [hunk 4]
```diff
@@ -56,7 +56,7 @@ public CartRemoveIngredientCommand parse(String args) throws ParseException {
      * Returns true if none of the prefixes contains empty {@code Optional} values in the given
      * {@code ArgumentMultimap}.
      */
-    private static boolean arePrefixesPresent(ArgumentMultimap argumentMultimap, Prefix... prefixes) {
+    static boolean arePrefixesPresent(ArgumentMultimap argumentMultimap, Prefix... prefixes) {
```
Comment (gohkh) [inline src/main/java/seedu/address/logic/parser/cart/CartRemoveIngredientCommandParser.java]: What is the reason for making this non-private? Is there another class using this method?

File: src/test/java/seedu/address/commons/core/ConfigTest.java [hunk 5]
```diff
@@ -20,7 +19,7 @@ public void toString_defaultObject_stringReturned() {
     public void equalsMethod() {
         Config defaultConfig = new Config();
         assertNotNull(defaultConfig);
-        assertTrue(defaultConfig.equals(defaultConfig));
+        assertEquals(defaultConfig, defaultConfig);
```
Comment (gohkh) [inline src/test/java/seedu/address/commons/core/ConfigTest.java:22]: What's the rationale behind this change?

File: src/main/java/seedu/address/logic/parser/cart/CartRemoveIngredientCommandParser.java (hunk 4 above)
Comment (teo-jun-xiong) [reply src/main/java/seedu/address/logic/parser/cart/CartRemoveIngredientCommandParser.java]: I used it in the tests

File: src/test/java/seedu/address/commons/core/ConfigTest.java (hunk 5 above)
Comment (teo-jun-xiong) [reply src/test/java/seedu/address/commons/core/ConfigTest.java:22]: It's the same test case, I changed it because Intellij showed a popup

File: src/main/java/seedu/address/logic/parser/cart/CartAddRecipeIngredientCommandParser.java (hunk 3 above)
Comment (teo-jun-xiong) [reply src/main/java/seedu/address/logic/parser/cart/CartAddRecipeIngredientCommandParser.java]: Changed it to ParseException.

File: src/main/java/seedu/address/logic/parser/cart/CartRemoveIngredientCommandParser.java (hunk 4 above)
Comment (teo-jun-xiong) [reply src/main/java/seedu/address/logic/parser/cart/CartRemoveIngredientCommandParser.java]: On another note, I think we can maybe make this a public static method in a central class, maybe CookingPapaParser? It seems its being repeated in a few classes

Comment (gohkh) [reply src/main/java/seedu/address/logic/parser/cart/CartRemoveIngredientCommandParser.java]: I think we don't have to test private methods? It doesn't seem like making these methods non-private for the sake of testing them is a good idea. The method would already be tested implicitly by testing the parsing, and the coverage in terms of lines executed would be the same. Also, testing all these methods seems quite inefficient in terms of uncovering new bugs that wouldn't also fail for the parsing methods.

Comment (teo-jun-xiong) [reply src/main/java/seedu/address/logic/parser/cart/CartRemoveIngredientCommandParser.java]: That makes sense. But regardless, it does seem like it is being repeated in 7 classes. Do you think we should move it to one location and make the 7 classes call that method?

Comment (gohkh) [reply src/main/java/seedu/address/logic/parser/cart/CartRemoveIngredientCommandParser.java]: Yeah, that's a good idea. Perhaps we could put it in the `ArgumentMultimap` class directly.

Comment (teo-jun-xiong) [reply src/main/java/seedu/address/logic/parser/cart/CartRemoveIngredientCommandParser.java]: Doing so inevitably requires us to make it public/package-private though

Comment (gohkh) [reply src/main/java/seedu/address/logic/parser/cart/CartRemoveIngredientCommandParser.java]: I think that's fine, because the `ArgumentMultimap` contains the prefix-argument mappings, and it seems to make sense to have a method where clients can check if certain prefixes are present, similar to `contains()` type of methods. We could even make it public, but at least we can cut down on the repeated code, which doesn't seem to fit very well in all the parser classes anyway.

**Review by gohkh — APPROVED** [review]
Thanks for adding all the tests
